projects/asn1-proto/asn1_proto_to_der.cc

```diff
@@ -0,0 +1,156 @@
+#include "asn1_proto_to_der.h"
+
+namespace asn1_proto {
+
+uint8_t ASN1ProtoToDER::GetNumBytes(const size_t num, const size_t base) {
+  uint8_t base_bits = log2(base);
+  for (uint8_t num_bits = sizeof(num) * 8; num_bits > base_bits;
+       num_bits -= base_bits) {
+    if ((num >> num_bits)) {
+      return ceil((double)num_bits / base_bits);
+    }
+  }
+  // Special-case: zero requires one, not zero bytes.
+  return 1;
+}
+
+void ASN1ProtoToDER::AppendVariableInt(const size_t value, const size_t pos) {
+  std::vector<uint8_t> len_vec;
+  for (uint8_t shift = GetNumBytes(value, 256); shift != 0; --shift) {
+    len_vec.push_back((value >> ((shift - 1) * 8)) & 0xFF);
+  }
+  encoder_.insert(encoder_.begin() + pos, len_vec.begin(), len_vec.end());
+}
+
+size_t ASN1ProtoToDER::EncodeOverrideLength(const std::string raw_len,
+                                            const size_t len_pos) {
+  encoder_.insert(encoder_.begin() + len_pos, raw_len.begin(), raw_len.end());
+  return raw_len.size();
+}
+
+size_t ASN1ProtoToDER::EncodeIndefiniteLength(const size_t len_pos) {
+  AppendVariableInt(0x80, len_pos);
+  // The PDU's value is from |len_pos| to the end of |encoder_|, so just add an
+  // EOC marker to the end.
+  AppendVariableInt(0x00, encoder_.size());
+  AppendVariableInt(0x00, encoder_.size());
+  return 3;
+}
+
+size_t ASN1ProtoToDER::EncodeDefiniteLength(const size_t actual_len,
+                                            const size_t len_pos) {
+  AppendVariableInt(actual_len, len_pos);
+  size_t len_num_bytes = GetNumBytes(actual_len, 256);
+  // X.690 (2015), 8.1.3.3: The long-form is used when the length is
+  // larger than 127.
+  // Note: |len_num_bytes| is not checked here, because it will return
+  // 1 for values [128..255], but those require the long-form length.
+  if (actual_len > 127) {
+    // See X.690 (2015) 8.1.3.5.
+    // Long-form length is encoded as a byte with the high-bit set to indicate
+    // the long-form, while the remaining bits indicate how many bytes are used
+    // to encode the length.
+    AppendVariableInt((0x80 | len_num_bytes), len_pos);
```

Comment:
Similarly, the rename helps highlight that we're not using a var int here:

```suggestion
    encoder_.insert(encoder_.begin() + len_pos, (0x80 | len_num_bytes));
```